### DISCUSSION TITLE
Add default response for all the endpoints with Microsoft.AspNetCore.OpenApi

### DISCUSSION THREAD

**Discussion by marcominerva** [dotnet/aspnetcore#59304] [Q&A]
Hi!

I would like to add a default **Error** response for all the endpoints of my Minimal API:

![image](https://github.com/user-attachments/assets/08ab2cf2-2dec-4517-b374-c1620655f413)

So, I have created the following Operation Transformer:

```csharp
internal class DefaultResponseOperationTransformer : IOpenApiOperationTransformer
{
    public Task TransformAsync(OpenApiOperation operation, OpenApiOperationTransformerContext context, CancellationToken cancellationToken)
    {
        var defaultResponse = new OpenApiResponse()
        {
            Description = "Error",
            Content = new Dictionary<string, OpenApiMediaType>
            {
                [MediaTypeNames.Application.ProblemJson] = new()
                {
                    Schema = new()
                    {
                        Reference = new()
                        {
                            Id = nameof(ProblemDetails),
                            Type = ReferenceType.Schema
                        }
                    }
                }
            }
        };

        operation.Responses.TryAdd("default", defaultResponse);
        return Task.CompletedTask;
    }
}
```

This works only if I have at least one endpoint that produces a `Problem` response, so that the `ProblemDetails` schema is inserted in OpenAPI definition:

```csharp
app.MapPost("/api/pople", () =>
{
    // ...
})
.ProducesProblem(StatusCodes.Status400BadRequest);
```

To avoid this requirement, with Swashbuckle I had defined a document filter to automatically add the corresponding schema:

```csharp
internal class ProblemDetailsDocumentFilter : IDocumentFilter
{
    public void Apply(OpenApiDocument swaggerDoc, DocumentFilterContext context)
        => context.SchemaGenerator.GenerateSchema(typeof(ProblemDetails), context.SchemaRepository);
}
```

But this is no longer possible with **Microsoft.AspNetCore.OpenApi**. So, in this case what is the correct solution?

**Comment by martincostello** [comment]
I'm not aware of a way to push an entirely new schema into the document as-is.

Maybe if you defined the schema inline (not as reference, because it will be pointing at something not defined) the later bits of the pipeline will notice it's re-used in multiple places and hoist it up to be a schema reference for you?

If that doesn't work, wrap all the operations in a group and define the default error like with `ProducesProblem()` there so it gets applied to all endpoints?

**Reply by marcominerva** [reply]
> 
> Maybe if you defined the schema inline (not as reference, because it will be pointing at something not defined) the later bits of the pipeline will notice it's re-used in multiple places and hoist it up to be a schema reference for you?
> 

@captainsafia could you tell me if `Microsoft.AspNetCore.OpenApi` contains such a logic to identify that the same schema is re-used multiple times?

> If that doesn't work, wrap all the operations in a group and define the default error like with `ProducesProblem()` there so it gets applied to all endpoints?

If the above approach does not work, I think this could be a viable solution.

**Reply by captainsafia** [reply]
> https://github.com/dotnet/aspnetcore/issues/58719

Yes, it does. You'll want to create the `OpenApiSchema` in your transformer. You'll also want to set the `x-schema-id` property in `OpenApiSchema.Annotations` to set the reference ID that should be used for it.

Also, I think what you've proposed here is covered by the issue that @mikekistler opened over in https://github.com/dotnet/aspnetcore/issues/58719.

It encapsulates updating the ApiExplorer layer to sets these error responses as the default response so that all consumer of it can use them and you don't have to rely on transformers and filters to do it. If that sounds interesting to you, go ahead and give the issue an upvote.

**Reply by marcominerva** [reply]
Thank you @captainsafia and @martincostello. I have tried to follow your suggestion, but the schema isn't reused (it is repeated for each endpoint). Can you take a look at https://github.com/marcominerva/TinyHelpers/blob/default_response/src/TinyHelpers.AspNetCore/OpenApi/DefaultResponseOperationTransformer.cs and help me finding what is wrong? Thank you very much 😄

**Reply by captainsafia** [reply]
@marcominerva By not reused, do you mean that you encounter duplicate schemas with numeric differentiators at the end? If you have a sample app I could run to debug, that would be helpful.

I don't see any reason for the schema comparison to fail if the same `defaultSchema` object is used in the transformer.

**Reply by marcominerva** [reply]
@captainsafia thank you for your support! I have created a small repro to showcase the issue: https://github.com/marcominerva/DefaultResponseTransformer

**Reply by captainsafia** [reply]
@marcominerva Eep -- this is my fault.  Setting the schema ID isn't sufficient since those get processed by the OpenApiSchemaStore which is an internal API.

I tried [this approach](https://github.com/captainsafia/tests/blob/503b6eadb4333b07f132a7783eca8dbbda7638b6/DefaultResponseTransformer/Program.cs#L12) to get the desired result by inserting the schema via a document generator. What do you think about this?

**Reply by marcominerva** [reply]
@captainsafia Yes, your approach correctly defines and reuses the `ProblemDetails` schema, thanks. But, unfortunately, there is a last issue. If I have an endpoint that already defines a _Problem_ response, for example:

```csharp
app.MapGet("/api/ping", () =>
{
    return TypedResults.Ok();
})
.ProducesProblem(StatusCodes.Status400BadRequest);
```

And then I use also your Transformers, because I still want the **default** response, I get an exception:

```
Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddleware[1]
      An unhandled exception has occurred while executing the request.
      System.ArgumentException: An item with the same key has already been added. Key: ProblemDetails
         at System.Collections.Generic.Dictionary`2.TryInsert(TKey key, TValue value, InsertionBehavior behavior)
         at System.Collections.Generic.Dictionary`2.Add(TKey key, TValue value)
         at Microsoft.AspNetCore.OpenApi.OpenApiSchemaReferenceTransformer.TransformAsync(OpenApiDocument document, OpenApiDocumentTransformerContext context, CancellationToken cancellationToken)
         at Microsoft.AspNetCore.OpenApi.OpenApiDocumentService.ApplyTransformersAsync(OpenApiDocument document, IServiceProvider scopedServiceProvider, CancellationToken cancellationToken)
         at Microsoft.AspNetCore.OpenApi.OpenApiDocumentService.GetOpenApiDocumentAsync(IServiceProvider scopedServiceProvider, CancellationToken cancellationToken)
         at Microsoft.AspNetCore.OpenApi.OpenApiDocumentService.GetOpenApiDocumentAsync(IServiceProvider scopedServiceProvider, CancellationToken cancellationToken)
         at Microsoft.AspNetCore.Builder.OpenApiEndpointRouteBuilderExtensions.<>c__DisplayClass0_0.<<MapOpenApi>b__0>d.MoveNext()
```

I see that in your code you use the `document.Components.Schemas.TryAdd` method, but when it is executed, the Schema collection is still empty, so it always adds the schema before the default implementation.

**Reply by UltraWelfare** [reply]
> @marcominerva Eep -- this is my fault. Setting the schema ID isn't sufficient since those get processed by the OpenApiSchemaStore which is an internal API.
> 
> I tried [this approach](https://github.com/captainsafia/tests/blob/503b6eadb4333b07f132a7783eca8dbbda7638b6/DefaultResponseTransformer/Program.cs#L12) to get the desired result by inserting the schema via a document generator. What do you think about this?

I've tried it and it seems to work (apart from @marcominerva's problem) but I wonder if there's a way to get the schema without manually typing it.

I have the same use-case of wanting to add a "400" response to all endpoints, however the class that I return also contains an enum.. I'm not sure how to go ahead and type this on `TransformAsync` of `DefaultResponseDocumentTransformer`.

**Reply by marcominerva** [reply]
@captainsafia I have managed to solve the duplication problem by explicitly checking if there already is an endpoint that returns a `ProblemDetails` response, so I add it to the schemas collection only if does not exist:

```csharp
app.MapGet("/api/ping", () =>
{
    return TypedResults.Ok();
})
.ProducesProblem(StatusCodes.Status400BadRequest);

public Task TransformAsync(OpenApiDocument document, OpenApiDocumentTransformerContext context, CancellationToken cancellationToken)
{
    // Checks if the ProblemDetails type is already defined in the document (because there is at least one endpoint that returns it).
    var isProblemDetailsSchemaDefined = context.DescriptionGroups
       .SelectMany(g => g.Items).SelectMany(i => i.SupportedResponseTypes)
       .Any(type => type.Type == typeof(ProblemDetails));

    if (isProblemDetailsSchemaDefined)
    {
        return Task.CompletedTask;
    }

    document.Components ??= new();
    document.Components.Schemas.TryAdd(nameof(ProblemDetails), new OpenApiSchema
    {
```

I have updated my sample at https://github.com/marcominerva/DefaultResponseTransformer with this workaround.
